### PR TITLE
WorkGroup Dispatch Policies

### PR DESCRIPTION
**Pull request by MrBurmark**
# Add Dispatch Policies to WorkGroup

This adds an additional policy to the WorkGroup construct that allows for different methods of dispatching work. Previously WorkGroup always used functions pointers to dispatch across from type erased to typed, for example when calling the callable object. This adds another indirect method using virtual functions. This also adds a "direct" dispatch method using a switch like construct to avoid an indirect function call and allow inlining and better compiler optimization. However this direct dispatch policy is less flexible than in indirect policy and requires upfront knowledge of all the range, callable pairs that may be enqueued.

The macro version of RAJA_ENABLE_HIP_INDIRECT_FUNCTION_CALL has been changed to only be defined starting with hip 5.1 instead of hip 4.3 as that is the first hip version that passes all of the tests using this feature. This cmake variable version of RAJA_ENABLE_HIP_INDIRECT_FUNCTION_CALL has been removed because it did not control the macro version and was redundant. The way we handle workgroup tests that are expected to failure has been changed. The failing cases of these tests are now disabled via template specialization when support for indirect function calls is not available based on the RAJA_ENABLE_HIP_INDIRECT_FUNCTION_CALL macro.

- This PR is a feature
- It does the following:
  - Modifies WorkGroup to add a dispatch policy
  - Adds direct dispatch at the request of Peter

# Design review (for API changes or additions---delete if unneeded)

On (date), we reviewed this PR.  We discussed the design ideas:

1. First idea or goal
2. Second idea
3. Third idea

This PR implements 1. and 3.  It leaves out 2. for the following reasons

- (impractical)
- (too big)
- (not a good idea anyway)

### CODE REVIEW

**Review by rhornung67 — APPROVED** [review]

**Comment by rhornung67** [comment]
@MrBurmark checks all green after I restarted the job that timed out earlier.  You may want to poke a few others to review this.

**Review by artv3 — APPROVED** [review]

**Comment by artv3** [comment]
Auto-merge! I didn't know we had that 👍

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -75,7 +75,7 @@ stages:
 
 .build_toss_4_x86_64_ib_corona_script:
   script:
-    - flux mini run --time-limit=30m --nodes=1 scripts/gitlab/build_and_test.sh
+    - flux mini run --time-limit=40m --nodes=1 scripts/gitlab/build_and_test.sh
```
Comment (adrienbernede) [inline .gitlab-ci.yml:78]:
```suggestion
    - flux mini run --time-limit=${DEFAULT_TIME}m --nodes=1 scripts/gitlab/build_and_test.sh
```

**Review by adrienbernede — COMMENTED** [review]
Suggestion

**Pull request merged**